// node-red-contrib-powerbi/powerbi-import.js
var request = require('request');

module.exports = function(RED) { 
	function PowerBIImport(config) { 
		RED.nodes.createNode(this,config);

		// Retrieve the connection node.
		this.connection = RED.nodes.getNode(config.connection);
		
		var node = this; 

		node.on('input', function(msg) {
/*
			request({
				method: 'POST', 
				url: 'https://api.powerbi.com/v1.0/mpsit/imports',
				headers: {
					'Content-Type': 'multipart/form-data; boundary=---BOUNDARY',
					'Authorization': 'Bearer tokenString'
				},
				body: "-----BOUNDARY
Content-Disposition: form-data; name=\"image[file]\"; filename=\"image.jpg\"
Content-Type: image/jpg
Content-Transfer-Encoding: base64

${msg.payload}
-----BOUNDARY"
			}, 
			function(error, response, body) {
				msg.status = response.statusCode; 
				msg.headers = JSON.stringify(response.headers); 
				msg.payload = body;

				*/
				node.send(msg);
			//});
		});
	}

	RED.nodes.registerType('powerbi-import', PowerBIImport, {
		settings: {
			powerbiImportKey: { 
				value: "1ec09691-148f-4548-ac09-9532a96cbf40", 
				exportable: true
			}
		}
	});
}
